File: Strivers-A2Z-DSA-Sheet/04.Linked List/1. Single Linked List/python/04.Count_the_number_of_nodes_linked_list.py
"""
QUESTION:
Given a singly linked list, find the length (number of nodes) of the linked list.

Example 1:
  Input:  1 -> 2 -> 3 -> 4 -> 5
  Output: 5

Example 2:
  Input:  2 -> 4 -> 6 -> 7 -> 5 -> 1 -> 0
  Output: 7

APPROACH:
1. Initialize a counter cnt = 0.
2. Traverse the list from head to None, incrementing cnt at each node.
3. Return cnt.

Time Complexity:  O(N) — visit every node once
Space Complexity: O(1)
"""


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


def get_count(head):
    cnt = 0
    curr = head
    while curr:
        cnt += 1
        curr = curr.next
    return cnt


# Helper: build linked list from array
def build_ll(arr):
    if not arr:
        return None
    head = Node(arr[0])
    curr = head
    for i in range(1, len(arr)):
        curr.next = Node(arr[i])
        curr = curr.next
    return head


# Driver code
if __name__ == "__main__":
    head1 = build_ll([1, 2, 3, 4, 5])
    print(get_count(head1))  # 5

    head2 = build_ll([2, 4, 6, 7, 5, 1, 0])
    print(get_count(head2))  # 7

File: Strivers-A2Z-DSA-Sheet/04.Linked List/1. Single Linked List/javascript/04.Count_the_number_of_nodes_linked_list.js
/*
QUESTION:
Given a singly linked list, find the length (number of nodes) of the linked list.

Example 1:
  Input:  1 -> 2 -> 3 -> 4 -> 5
  Output: 5

Example 2:
  Input:  2 -> 4 -> 6 -> 7 -> 5 -> 1 -> 0
  Output: 7

APPROACH:
1. Initialize a counter cnt = 0.
2. Traverse the list from head to null, incrementing cnt at each node.
3. Return cnt.

Time Complexity:  O(N) — visit every node once
Space Complexity: O(1)
*/

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

function getCount(head) {
    let cnt = 0;
    let curr = head;
    while (curr) {
        cnt++;
        curr = curr.next;
    }
    return cnt;
}

// Helper: build linked list from array
function buildLL(arr) {
    if (!arr.length) return null;
    const head = new Node(arr[0]);
    let curr = head;
    for (let i = 1; i < arr.length; i++) {
        curr.next = new Node(arr[i]); curr = curr.next;
    }
    return head;
}

// Driver code
const head1 = buildLL([1, 2, 3, 4, 5]);
console.log(getCount(head1)); // 5

const head2 = buildLL([2, 4, 6, 7, 5, 1, 0]);
console.log(getCount(head2)); // 7

